lib/redlock/client.rb

```diff
@@ -169,15 +169,17 @@ def initialize(connection)
       def lock(resource, val, ttl, allow_new_lock)
         recover_from_script_flush do
           @redis.with { |conn|
-            conn.call('EVALSHA', Scripts::LOCK_SCRIPT_SHA, 1, resource, val, ttl, allow_new_lock)
+            # NOTE: is idempotent and safe to retry
+            conn.call_once('EVALSHA', Scripts::LOCK_SCRIPT_SHA, 1, resource, val, ttl, allow_new_lock)
```

Comment:
I assume you mean unsafe if you're using `call_once`?

Reply:
ah, comment got out of date with my code.  I changed my mind that locking probably isn't idempotent and change the code from `call` to `call_once` didn't update the comment.

Reply:
now the logic and comments are in sync

unlock/lock

```ruby
# NOTE: is not idempotent and unsafe to retry
conn.call_once(...)
```

load_scripts/get remaining ttl

```ruby
# NOTE: is idempotent and safe to retry
conn.call(...)
```

aside, interesting logic inside where it defines `redis(&block)`  https://github.com/sidekiq/sidekiq/blob/v7.0.5/lib/sidekiq/config.rb#L118-L133

```ruby
    def redis
      raise ArgumentError, "requires a block" unless block_given?
      redis_pool.with do |conn|
        retryable = true
        begin
          yield conn
        rescue RedisClientAdapter::BaseError => ex
          # 2550 Failover can cause the server to become a replica, need
          # to disconnect and reopen the socket to get back to the primary.
          # 4495 Use the same logic if we have a "Not enough replicas" error from the primary
          # 4985 Use the same logic when a blocking command is force-unblocked
          # The same retry logic is also used in client.rb
          if retryable && ex.message =~ /READONLY|NOREPLICAS|UNBLOCKED/
            conn.close
            retryable = false
            retry
          end
          raise
        end
      end
    end
```